src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets

```diff
@@ -861,13 +861,25 @@ Copyright (c) .NET Foundation. All rights reserved.
       <PublishedSingleFileName>$(AssemblyName)$(_NativeExecutableExtension)</PublishedSingleFileName>
       <PublishedSingleFilePath>$(PublishDir)$(PublishedSingleFileName)</PublishedSingleFilePath>
     </PropertyGroup>
-
   </Target>
 
+  <Target Name="BeforeBundle"
+      DependsOnTargets="_ComputeFilesToBundle"
+      Condition="'$(PublishSingleFile)' == 'true'">
+
+    <ItemGroup>
+      <FilesToBundle Include="@(_FilesToBundle)"/>
+    </ItemGroup>
+
+    <PropertyGroup>
+      <AppHostFile>$(PublishedSingleFileName)</AppHostFile>
+    </PropertyGroup>
+  </Target>
+  
   <UsingTask TaskName="GenerateBundle" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
   <Target Name="GenerateSingleFileBundle"
           Condition="'$(PublishSingleFile)' == 'true'"
-          DependsOnTargets="_ComputeFilesToBundle"
+          DependsOnTargets="_ComputeFilesToBundle;BeforeBundle"
```

Comment:
What's the intended consumption pattern here? Add a Target that is before `GenerateSingleFileBundle` and depends on `BeforeBundle`? Would something like `GenerateSingleFileBundle` depending on `$(GenerateSingleFileBundleDependsOn)` and have it just be `_ComputeFilesToBundle;BeforeBundle` + user provided targets where they can transform the itemgroup as needed? @rainersigwald, what's the most idiomatic/user safe way you'd recommend for this extensibility?

Reply:
Something a little like https://github.com/dotnet/msbuild/blob/HEAD/src/Tasks/Microsoft.Common.CurrentVersion.targets#L871-L883

Reply:
I agree that we should also define the way to "plugin". For build it's enough to just define a target with the right name `BeforeBuild` - maybe we should do the same.

Reply:
Common targets generally use this pattern:

1. Target Foo `DependsOnTargets="$(FooDependsOn)"`
2. The default `FooDependsOn` is `BeforeFoo;CoreFoo;AfterFoo`
3. Stub implementations of `BeforeFoo` and `AfterFoo` that do nothing are defined.
4. Users may override the definitions of `BeforeFoo` or `AfterFoo` with customizations.
5. Users may also define their own targets that hook to any of those targets with `BeforeTargets` or `AfterTargets`.

The "override" extension strategy was recommended and used in many projects, but is harder to deal with in SDK-style projects, because the last definition of a target in MSBuild's expanded project wins, and a stub target defined in SDK/common .targets will override a target of the same name defined in the project file. As a result, **we now recommend creating a new target and hooking to the appropriate target** over the override approach. That also lets you give the target a meaningful name (like `SignEmbeddedPayload`).

Here, I don't think `AfterBundle` makes much sense so just adding `BeforeBundle` seems reasonable. I wonder though if it might be better to rename this target `PrepareForBundle` and document something like:

```xml
<Target Name="SignBundledFile" BeforeTargets="GenerateSingleFileBundle" DependsOnTargets="PrepareForBundle">
```

?

That's basically what the linked project is doing, the "bad" thing they're doing is modifying the "private" `@(_FilesToBundle)`. Here you're making that "public".

Reply:
So the suggestions are, if I get it right:
- `GenerateSingleFileBundle` should use `FilesToBundle`, not internal `_FilesToBundle`.
I had small concern about overrider adding/removing items to the list. Changing items is basically the purpose here, but I was not sure whether changing the list itself.
Anyways, if it seems preferable to use the same `FilesToBundle`, I will change to that.
- Rename `BeforeBundle` --> `PrepareForBundle`.
I do not have an opinion here. Either seem ok to me. Just want to be sure that people more invested in the name of the target are in agreement.
- and test this somehow.

Is this right?

Reply:
> * `GenerateSingleFileBundle` should use `FilesToBundle`, not internal `_FilesToBundle`.
>   I had small concern about overrider adding/removing items to the list. Changing items is basically the purpose here, but I was not sure whether changing the list itself.

Yeah, it's an odd thing, but necessary. The user might deposit the signed files elsewhere. So yeah, the inputs and implementation need to use the exposed ItemGroup.

> * Rename `BeforeBundle` --> `PrepareForBundle`.

I am happy with either name. Maybe we should add the example to the design, say before `GenerateSingleFileBundle` and depend on the name we chose.

Reply:
And then document that the way to plug in is to do what @rainersigwald suggested:

```XML
<Target Name="MySignBundledFile" BeforeTargets="GenerateSingleFileBundle" DependsOnTargets="PrepareForBundle">
```

Reply:
> Yeah, it's an odd thing, but necessary. The user might deposit the signed files elsewhere.

Right, if files are outside of the build's control and effectively or actually read-only. 
I assumed that changing the items metadata would be sufficient to repoint to the new location, but I am probably wrong about that and some scenarios will call for having access to the actual list.

Reply:
I have noticed a small advantage in calling this `PrepareForBundle` - so that noone would ask "why there is no corresponding `AfterBundle` ?".

So I will go with `PrepareForBundle` then.